node/src/components/storage.rs

```diff
@@ -1616,86 +1579,194 @@ fn initialize_block_body_v1_db(
         tx,
         block_body_v2_db,
-        transfer_hashes_db,
-        digest_key_to_delete,
-    )?;
-    // Don't delete proposer since it might be shared
-    Ok(())
+        proposer_db,
+        transfer_hashes_with_proof.merkle_proof_of_rest(),
+    )? {
+        Some(proposer_with_proof) => {
+            debug_assert_eq!(*proposer_with_proof.merkle_proof_of_rest(), hash::SENTINEL1);
+            proposer_with_proof
+        }
+        None => return Ok(None),
+    };
+    let block_body = BlockBody::new(
+        proposer_with_proof.take_value(),
+        deploy_hashes_with_proof.take_value(),
+        transfer_hashes_with_proof.take_value(),
+    );
+
+    Ok(Some(block_body))
 }
 
 /// Purges stale entries from the (Merklized) block body database, and checks the integrity of the
 /// remainder if `should_check_integrity` is true.
 fn initialize_block_body_v2_db(
     env: &Environment,
+    block_header_db: &Database,
     block_body_v2_db: &Database,
     deploy_hashes_db: &Database,
     transfer_hashes_db: &Database,
-    _proposer_db: &Database,
-    deleted_block_body_hashes: HashSet<Digest>,
-    // should_check_integrity: bool,
+    proposer_db: &Database,
+    should_check_integrity: bool,
 ) -> Result<(), LmdbExtError> {
     info!("initializing v2 block body database");
+
     let mut txn = env.begin_rw_txn()?;
-    for digest_key_to_delete in deleted_block_body_hashes {
-        info!(?digest_key_to_delete, "deleting v2 block body");
-        delete_single_block_body_v2(
-            &mut txn,
-            block_body_v2_db,
-            deploy_hashes_db,
-            transfer_hashes_db,
-            digest_key_to_delete,
-        )?;
+
+    let block_body_hash_to_header_map =
+        construct_block_body_to_block_header_reverse_lookup(&txn, block_header_db)?;
+
+    // This will store all the keys that are reachable from a block header, in all the parts
+    // databases (we're basically doing a mark-and-sweep below).
+    let mut live_digests = HashSet::new();
+
+    for (body_hash, header) in &block_body_hash_to_header_map {
+        if header.hashing_algorithm_version() != HashingAlgorithmVersion::V2 {
+            // We're only interested in body hashes for V2 blocks here
+            continue;
+        }
+        let mut current_digest = *body_hash;
+        loop {
+            live_digests.insert(current_digest);
+            let [key_to_part_db, merkle_proof_of_rest]: [Digest; 2] =
+                match txn.get_value(*block_body_v2_db, &current_digest)? {
+                    Some(slice) => slice,
+                    None => {
+                        return Err(CouldNotFindBlockBodyPart {
+                            merkle_linked_list_node_hash: current_digest,
+                        })
+                    }
+                };
+            live_digests.insert(key_to_part_db);
+            current_digest = merkle_proof_of_rest;
+            if current_digest == hash::SENTINEL1 {
```

Comment:
We could also `break` if `live_digests.contains(current_digest)`, I think, to avoid redundant lookups?
(And we could add `SENTINEL1` to that set, so we only need to check `contains`.)

Reply:
Good idea :+1:

Reply:
(Although, we probably _would_ want to delete an entry with key `SENTINEL1` in theory. So, not sure about whether we want to add that to `live_digests`.)